c_src/myhtml_worker.c

```diff
@@ -256,140 +257,119 @@ read_parse_flags(ETERM* list)
+
+      continue;
+    } else {
+      if ((myhtml_node_is_close_self(current_node)  || myhtml_node_is_void_element(current_node))
+      && (*parse_flags & FLAG_NIL_SELF_CLOSING)) {
+        children = prefab->atom_nil;
+      } else {
+        children = prefab->empty_list;
+      }
     }
 
-    // attributes
-    attrs = build_node_attrs(prefab, tree, node);
+    myhtml_tag_id_t tag_id = myhtml_node_tag_id(current_node);
+    myhtml_namespace_t tag_ns = myhtml_node_namespace(current_node);
 
-    // children
-    children = build_node_children(prefab, tree, node, parse_flags);
+    if (tag_id == MyHTML_TAG__TEXT)
+    {
+      size_t text_len;
 
-    if (!(*parse_flags & FLAG_HTML_ATOMS) || (tag_id == MyHTML_TAG__UNDEF || tag_id == MyHTML_TAG_LAST_ENTRY || tag_ns != MyHTML_NAMESPACE_HTML))
+      const char* node_text = myhtml_node_text(current_node, &text_len);
+      result = erl_mk_binary(node_text, text_len);
+    }
+    else if (tag_id == MyHTML_TAG__COMMENT)
     {
-      tag = erl_mk_binary(tag_string, tag_string_len);
-      /* ETERM* tuple3[] = {tag, attrs, children}; */
-      /* result = erl_mk_tuple(tuple3, 3); */
-      result = erl_format("{~w, ~w, ~w}", tag, attrs, children);
+      size_t comment_len;
+      const char* node_comment = myhtml_node_text(current_node, &comment_len);
+
+      // For <!----> myhtml_node_text will return a null pointer, which will make erl_format segfault
+      ETERM* comment = erl_mk_binary(node_comment ? node_comment : "", comment_len);
+
+      if (*parse_flags & FLAG_COMMENT_TUPLE3)
+      {
+        result = erl_format("{comment, [], ~w}", comment);
+      }
+      else
+      {
+        result = erl_format("{comment, ~w}", comment);
+      }
     }
     else
     {
-      // tag = erl_mk_atom(tag_string);
-      tag = erl_mk_atom(tag_string);
-      /* ETERM* tuple3[] = {tag, attrs, children}; */
-      /* result = erl_mk_tuple(tuple3, 3); */
-      result = erl_format("{~w, ~w, ~w}", tag, attrs, children);
-    }
+      ETERM* tag;
+      ETERM* attrs;
+
+      // get name of tag
+      size_t tag_name_len;
+      const char *tag_name = myhtml_tag_name_by_id(tree, tag_id, &tag_name_len);
+      // get namespace of tag
+      size_t tag_ns_len;
+      const char *tag_ns_name_ptr = myhtml_namespace_name_by_id(tag_ns, &tag_ns_len);
+      char buffer [tag_ns_len + tag_name_len + 2];
+      char *tag_string = buffer;
+      size_t tag_string_len;
+
+      if (tag_ns != MyHTML_NAMESPACE_HTML)
+      {
+        // tag_ns_name_ptr is unmodifyable, copy it in our tag_ns_buffer to make it modifyable.
+        char tag_ns_buffer[tag_ns_len];
+        strcpy(tag_ns_buffer, tag_ns_name_ptr);
+        lowercase(tag_ns_buffer);
+
+        tag_string_len = tag_ns_len + tag_name_len + 1; // +1 for colon
+	// +1 for the null-byte, which we won't use because erlang binary strings are not null-terminated
+	snprintf(tag_string, tag_string_len + 1, "%s:%s", tag_ns_buffer, tag_name);
```

Comment:
use `sizeof(tag_string)` for this.  always use sizeof() for string function boundaries.

Reply:
Using `sizeof(tag_string)` had a weird bug where it would empty the n-th char of the tag, where n is the length of the namespace. `sizeof(buffer)` worked fine though, not sure why since they point to the same thing

Reply:
oh yeah, it should be sizeof(buffer) actually, because tag_string is a pointer.  I misread the code originally :)